### ISSUE TITLE
[idea] implementing rate limiting per queue?

### ISSUE THREAD

**Issue by 34code** [basecamp/solid_queue#97]
Great release! Looking forward to trying this in tandem with Mission Control in 2024!

I had a question about `concurrency` and how it might play with `uniqueness` and `rate-limits` for things like external APIs..

Whats the recommended way for dealing with external APIs which impose a rate limit (e.g. 5 requests per second) with solid_queue?

The concurrency controls with `key` is great btw, I'm often concerned about two workers writing to the same record concurrently so this is super helpful :)

**Comment by rosa** [comment]
Hey @34code, thank you so much! 🙏 

Great question! For rate-limited external API calls, we've usually relied on the number of workers and threads running jobs (we have a `throttled` queue that we run with very few workers), and on `retry_on` with polynomial backoff or with a custom time to wait depending on the rate limit and the response that made us retry. Using concurrency controls with a `key` might work, too, depending on how fast the jobs are to execute and on your load in general (how many jobs you're running of that type). 

In general, for high-volume jobs, I tend to avoid concurrency controls if I can because they add overhead, having to gain the lock to run and then unblock other waiting jobs. We're currently using concurrency controls for a high-volume job, and that led us to https://github.com/basecamp/solid_queue/pull/99 😅 

Sorry that I don't have a more definitive answer! In these cases, what I do is look at the numbers (how many jobs per second, the rate limit imposed by external APIs, how long the jobs take, and what's an acceptable latency for these jobs to have, that's it, is it ok if they take longer to run because we're throttling them and it doesn't matter if we throttle too aggressively, or are they user-facing and we need to run them as fast as we can within limits?) and then decide based on that 😅

**Comment by 34code** [comment]
got it, thanks for the tips! currently im on sidekiq and run with a concurrency of 1 in sidekiq config.. and some sleep in the jobs like 0.7s .. this allows around 1 job per second to run, and my rate limit is 60 rps. It sounds like i should simply be able to recreate this scenario with `solid_queue`?

**Comment by rosa** [comment]
@34code yes, I think so. Let me know if you have any problems with it!

**Comment by Alagaesia93** [comment]
Jumping on this discussion, but happy to open another issue, if needed!

Your approach @rosa works well for time based requests, but Shopify for example uses the leaky bucket algorithm, with points. Sidekiq ent has a fantastic solution for this, in fact most of the Shopify app devs use it.

How would you approach this problem? 
On a separate note, should there be something in the readme about ideas on how to rate limit? happy to contribute!

**Comment by rosa** [comment]
Hey @Alagaesia93, thanks for this! I didn't know this about Shopify, as I've never used Shopify APIs, so this is fantastic information. I'll reopen this to continue the discussion, although I need to think carefully about this one so it might take me a few days to have an answer as I have to work on other stuff this week 😅 

About rate-limiting in the README, sure! Feel free to open a PR if you want, with something to start. More than specific info on rate-limiting and APIs, it might make sense to approach it as a way to throttle jobs, in a more generic way 🤔

**Comment by Alagaesia93** [comment]
Hi @rosa! By any chance, did you get any time to think about this?

**Comment by rosa** [comment]
Hey @Alagaesia93, sorry for the delay. I did think about this, yes, and I'd approach a rate-limit like Shopify's externally to the jobs backend itself. That's it, it'd be something that runs and keeps track of the current state outside Solid Queue, and you'd use that either to schedule your jobs according to the rate-limit, or to do nothing or skip processing within the job if that'd mean you'd go over the rate-limit. I think this way of rate-limiting is orthogonal to the Active Job backend, and should live outside of it, so you could apply it to other components 🤔
